### ISSUE TITLE
permissions in Role should probably be represented with an Integral type

### ISSUE THREAD

**Issue by Annwan** [discord-haskell/discord-haskell#163]
Currently `rolePerms` in `Role` are of type `Data.Text.Text` but it actually is an integer representing a bitset, it would therefore make more sense to have it be an Integral type, it would also simplify permission calculations.

**Comment by L0neGamer** [comment]
Ideally we would have a type that it's easy to request permissions from without the library user having to have the bits required to hand

**Comment by XanderDJ** [comment]
I don't mind doing this

**Comment by Annwan** [comment]
uh something like
```hs
newtype Permission = Permission
    { permissionCreateInstantInvite :: Bool
    , permissionKickMembers :: Bool
    , permissionBanMembers :: Bool
    , permissionAdministrator :: Bool
    , -- ...
    }
```?

**Comment by XanderDJ** [comment]
I was more so thinking like:
`
data Permissions = CreateInstantInvite | ...

hasPermissions :: [Permission] -> (type RolePerms = Int) -> Bool
`
You could also check every permission to see if it is set and then make that type.

**Comment by Annwan** [comment]
That could work too, the only problem I might see with that approach is when setting permissions
If you want to update a permission with the record approach you just have to do something like that `newPerms = oldPerms {permissionSendMessage = False}` instead of having to filter the list (for deletion) or check against duplicates (when authorizing)

**Comment by XanderDJ** [comment]
```haskell
addPermissions, removePermissions :: [Permission] -> RolePerms -> RolePerms
newPermissions :: [Permission] -> RolePerms
```

I don't think that's as big of a problem if you just make sure the functions are there.

**Comment by L0neGamer** [comment]
Alternatively, we store a fixed size int for the permissions, then give functions which get the permissions from it and to it. This means that if they add permissions and we don't update, people can still access those permissions if they need.

**Comment by XanderDJ** [comment]
Well in my example RolePerms would just be a type alias for an integer so that would already be a thing.

**Comment by L0neGamer** [comment]
I think newtype would be preferred but yeah that sounds decent.

**Comment by Annwan** [comment]
yeah that'd work then
we'd just need to add
```hs
setPermission :: Permission -> RolePerms -> RolePerms
clearPermission :: Permission -> RolePerms -> RolePerms
```
also to be consitent with the way wrapper types are done in the lib it would be something more like
```hs
newtype RolePerms = RolePerms Integer
```

**Comment by XanderDJ** [comment]
Yea I get the newtype. I will get this done monday if that is fine.

**Comment by Annwan** [comment]
no problem, this is not urgent, the current system works, it's just a nice to have

**Comment by L0neGamer** [comment]
Closed by #164
